File: LeetCode_Algorithm/344(Reverse String)/344.py
from typing import List

class Solution:
    def reverseString(self, s: List[str]) -> None:
        # sol1 O(n), O(1) 296~370ms 34~12%
        # for i in range(len(s)//2):
        #     s[i], s[len(s)-i-1] = s[len(s)-i-1], s[i]
        # print(s)

        # sol2 217~284ms 73~39%
        # oneline
        # s.reverse()
        # print(s)

        # sol3 208~218ms 82~73%
        # recursive
        # def swap(left, right):
        #     if left < right:
        #         s[left], s[right] = s[right], s[left]
        #         swap(left+1, right-1)
        # swap(0, len(s)-1)
        # print(s)

        # sol4
        # iterative
        left = 0; right = len(s)-1
        while left < right:
            s[left], s[right] = s[right], s[left]
            left, right = left+1, right-1
        print(s)

print(Solution().reverseString(["h","e","l","l","o"]))
        

File: LeetCode_Algorithm/344(Reverse String)/344.js
var reverseString = function(s) {
  // sol2 119~154ms 58~22%
  // s.reverse()
  // console.log(s)

  // sol3 88~175ms 93~11%
  // recursive
  var swap = function(left, right){
    if (left < right){
      const temp = s[left]
      s[left] = s[right]; s[right] = temp
      swap(left+1, right-1)
    }
  }
  swap(0, s.length-1)
  console.log(s)

  // sol4 126~138ms 50~37%
  // var left = 0; 
  // var right = s.length-1
  // while (left < right) {
  //   const temp = s[left]
  //   s[left] = s[right]; s[right] = temp
  //   left = left+1; right = right-1
  // }
  // console.log(s)

};

reverseString(["h","e","l","l","o"])